Firestore/core/test/firebase/firestore/local/index_free_query_engine_test.cc

```diff
@@ -0,0 +1,336 @@
+              since_read_time != SnapshotVersion::None());
+    return LocalDocumentsView::GetDocumentsMatchingQuery(query,
+                                                         since_read_time);
+  }
+
+  void ExpectIndexFreeExecution(bool index_free) {
+    expect_index_free_execution = index_free;
+  }
+
+ private:
+  absl::optional<bool> expect_index_free_execution;
+};
+
+class IndexFreeQueryEngineTest : public ::testing::Test {
+ public:
+  IndexFreeQueryEngineTest()
+      : persistence_(MemoryPersistence::WithEagerGarbageCollector()),
+        remote_document_cache_(persistence_->remote_document_cache()),
+        query_cache_(persistence_->query_cache()),
+        index_manager_(new MemoryIndexManager()),
+        local_documents_view_(
+            remote_document_cache_,
+            persistence_->GetMutationQueueForUser(User::Unauthenticated()),
+            index_manager_.get()) {
+    query_engine_.SetLocalDocumentsView(&local_documents_view_);
+  }
+
+  /** Adds the provided documents to the query target mapping. */
+  void PersistQueryMapping(const std::vector<DocumentKey>& keys) {
+    persistence_->Run("PersistQueryMapping", [this, &keys]() {
+      DocumentKeySet remote_keys;
+      for (const DocumentKey& key : keys) {
+        remote_keys = remote_keys.insert(key);
+      }
+      query_cache_->AddMatchingKeys(remote_keys, TEST_TARGET_ID);
+    });
+  }
+
+  /** Adds the provided documents to the remote document cache. */
+  void AddDocuments(const std::vector<Document>& docs) {
+    persistence_->Run("AddDocuments", [this, &docs]() {
+      for (const Document& doc : docs) {
+        remote_document_cache_->Add(doc, doc.version());
+      }
+    });
+  }
+
+  DocumentSet ExpectIndexFreeQuery(const std::function<DocumentSet(void)>& f) {
+    local_documents_view_.ExpectIndexFreeExecution(true);
+    return f();
+  }
+
+  DocumentSet ExpectFullCollectionQuery(
+      const std::function<DocumentSet(void)>& f) {
+    local_documents_view_.ExpectIndexFreeExecution(false);
+    return f();
+  }
+
+  DocumentSet RunQuery(
+      const Query& query,
+      const SnapshotVersion& last_limbo_free_snapshot_version) {
+    const DocumentKeySet& remote_keys =
+        query_cache_->GetMatchingKeys(TEST_TARGET_ID);
+    DocumentMap docs = query_engine_.GetDocumentsMatchingQuery(
+        query, last_limbo_free_snapshot_version, remote_keys);
+    View view(query, DocumentKeySet());
+    ViewDocumentChanges viewDocChanges =
+        view.ComputeDocumentChanges(docs.underlying_map(), {});
+    return view.ApplyChanges(viewDocChanges).snapshot()->documents();
+  }
+
+ private:
+  std::unique_ptr<Persistence> persistence_;
+  RemoteDocumentCache* remote_document_cache_;
+  QueryCache* query_cache_;
+  std::unique_ptr<MemoryIndexManager> index_manager_;
+  IndexFreeQueryEngine query_engine_;
+  TestLocalDocumentsView local_documents_view_;
+};
+
+TEST_F(IndexFreeQueryEngineTest, UsesTargetMappingForInitialView) {
+  Query query =
```

Comment:
I updated this file to use `core::Query` and `Query` rather than `Query` and `Query(Resource())`. I actually think I liked the previous version more, but not enough to undo this change.

Reply:
I have to agree that the gains from using `testutil::Query` aren't that great. I guess the main argument in favor of using it would be consistency with existing tests.